pkg/controller/intrusiondetection/intrusiondetection_controller.go

```diff
@@ -550,9 +594,20 @@ func (r *ReconcileIntrusionDetection) setDefaultsOnIntrusionDetection(ctx contex
 			},
 		}
 
-		if err := r.client.Update(ctx, ids); err != nil {
-			return err
+	}
+
+	if len(ids.Spec.AnomalyDetection.StorageType) == 0 {
```

Comment:
This can result in an invalid configuration. e.g., if the user gives:

```
storageType: ""
storageClassName: my-storage-class
```

We'll default it to:

```
storageType: Ephemeral
storageClassName: my-storage-class
```

Which is not a valid config. I think we need to default this field conditionally based on whether a storage class name is given as well.

Reply:
think it would default to just `storageType: Ephemeral` since we replace the whole AnomalyDetection struct below

Reply:
That's right, but that in itself is a bug - our defaulting logic should never remove a field that was set by the user.

Reply:
understood, https://github.com/tigera/operator/pull/2094/commits/0242a6e687adf0bac93f81213982ec3928c81dd8 - ended up validating before default setting of the fields for ADSpec in case we see an ADSpec provided with that scenario 

and then check only to set default in the case if the whole ADSpec struct is empty

Reply:
That's somewhat atypical for the operator codebase, and makes it harder to understand this defaulting logic because it now needs to be read in the context of some prior validation. We want our validation and defaulting functions to make sense without requiring an understanding of each other.

I think we should go with my original suggestion and default this field conditionally:

```go
if len(storageType) == 0 {
  if (storageClass) == 0 {
    // Set to ephemeral.
  } else {
    // User specified a storage class, set to persistent
  }
}
```

Alternatively, we make this field required on the API and reject it unless the user specifies it. But that's a breaking change so probably not a valid option.

Reply:
https://github.com/tigera/operator/pull/2094/commits/fe81c4cb19794ca1b97dd65394e9276d1e3a623d